{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "R7gC4CcvaQ_R"
      },
      "source": [
        "## Frustratometer in Python\n",
        "<a href=\"https://colab.research.google.com/github/engelberger/frustrapy/blob/main/FrustraPy_colab.ipynb\">\n",
        "  <img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/>\n",
        "</a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_Ypytx4baQ_T"
      },
      "source": [
        "## Introduction\n",
        "\n",
        "The concept of frustration in proteins refers to the presence of conflicting interactions within the protein structure. These conflicts arise when the local interactions within a protein are not optimally stabilizing, leading to a certain degree of energetic frustration. The following is a summary please refer to the original paper for more details:\n",
        "\n",
        "[Protein Frustratometer 2: a tool to localize energetic frustration in protein molecules, now with electrostatics](https://academic.oup.com/nar/article/44/W1/W356/2499321)\n",
        "[FrustrometerR: an R-package to compute local frustration in protein structures, point mutants and MD simulations](https://academic.oup.com/bioinformatics/article/37/18/3038/6171179)\n",
        "\n",
        "\n",
        "## There are three main types of frustration\n",
        "\n",
        "* *Highly frustrated*: Highly frustrated regions in a protein are those where the local interactions are significantly destabilizing compared to what would be expected in an idealized, energetically minimized structure. These regions often play crucial roles in protein function, such as binding sites, allosteric sites, or regions involved in conformational changes. For example, in an enzyme, the active site might be highly frustrated to allow for substrate binding and catalysis, which require a certain degree of flexibility and adaptability.\n",
        "* *Neutral*: Neutral regions in a protein are those where the local interactions are neither significantly stabilizing nor destabilizing. These regions may not directly contribute to protein function but are essential for maintaining the overall structural integrity of the protein. Neutral regions can serve as a buffer between highly frustrated and minimally frustrated regions, allowing for the necessary flexibility and stability balance within the protein.\n",
        "* *Minimally frustrated*: Minimally frustrated regions in a protein are those where the local interactions are highly optimized and stabilizing. These regions typically form the stable core of the protein and are essential for maintaining the native folded state. Minimally frustrated regions often consist of hydrophobic residues that pack tightly together, forming a stable foundation for the protein structure. For example, in the case of globular proteins, the hydrophobic core is usually minimally frustrated, contributing to the overall stability of the folded state.\n",
        "\n",
        "## Significance of Frustration in Proteins:\n",
        "\n",
        "* Protein folding: During the protein folding process, the polypeptide chain navigates through an energy landscape to reach its native state. The concept of minimal frustration suggests that evolution has optimized the folding landscape to minimize energetic conflicts, allowing proteins to fold efficiently and avoid getting trapped in non-native states.\n",
        "* Allostery: Allosteric regulation in proteins often involves highly frustrated regions that undergo conformational changes upon ligand binding or other perturbations. These frustrated regions allow for the propagation of allosteric signals throughout the protein structure, enabling long-range communication and regulation of protein function.\n",
        "* Protein-protein interactions: Protein interfaces often contain a mix of highly frustrated and minimally frustrated regions. The highly frustrated regions may contribute to the specificity and adaptability of the interaction, while the minimally frustrated regions provide stability to the complex. The balance between frustration and stability at the interface is crucial for the formation and regulation of protein complexes."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "s-5c4wj0aQ_T"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Note: you may need to restart the kernel to use updated packages.\n"
          ]
        }
      ],
      "source": [
        "# @title Install\n",
        "%pip install -q biopython igraph leidenalg\n",
        "%pip install -q -U kaleido pandas scikit-learn matplotlib seaborn plotly\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "D0_Qe3uGaQ_U"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "2357.09s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
            "2364.05s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
            "2371.41s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
            "2378.37s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n",
            "2383.58s - pydevd: Sending message related to process being replaced timed-out after 5 seconds\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "The results directory /home/ceramirez/github/frustrapy/Results_example has been created.\n",
            "-----------------------------Filtering-----------------------------\n",
            "/home/ceramirez/github/frustrapy/Results_example/1fhj.done/1fhj.pdb\n",
            "1fhj.pdb\n",
            "/home/ceramirez/github/frustrapy/Results_example/1fhj.done/1fhj.pdb_equivalences.txt\n",
            "HI\n",
            "/home/ceramirez/github/frustrapy/Results_example/1fhj.done/1fhj.pdb\n",
            "1fhj.pdb\n",
            "/home/ceramirez/github/frustrapy/Results_example/1fhj.done/1fhj.pdb_equivalences.txt\n",
            "-----------------------------Preparing files-----------------------------\n",
            "Pdb file: 1fhj\n",
            "Output file: 1fhj\n",
            "-----------------------------Setting options-----------------------------\n",
            "-----------------------------Calculating-----------------------------\n",
            "LAMMPS (3 Mar 2020)\n",
            "Reading data file ...\n",
            "  orthogonal box = (-20000 -20000 -20000) to (20000 20000 20000)\n",
            "  1 by 1 by 1 MPI processor grid\n",
            "  reading atoms ...\n",
            "  1722 atoms\n",
            "  scanning bonds ...\n",
            "  3 = max bonds/atom\n",
            "  reading bonds ...\n",
            "  2288 bonds\n",
            "Finding 1-2 1-3 1-4 neighbors ...\n",
            "  special bond factors lj:   0          1          1         \n",
            "  special bond factors coul: 0          1          1         \n",
            "  5 = max # of 1-2 neighbors\n",
            "  5 = max # of special neighbors\n",
            "  special bonds CPU = 0.000322819 secs\n",
            "  read_data CPU = 0.00427699 secs\n",
            "574 atoms in group alpha_carbons\n",
            "574 atoms in group beta_atoms\n",
            "574 atoms in group oxygens\n",
            "ABC flag on\n",
            "Water flag on\n",
            "Burial flag on\n",
            "Tertiary_Frustratometer flag on\n",
            "\n",
            "Neighbor list info ...\n",
            "  update every 1 steps, delay 5 steps, check yes\n",
            "  max neighbors/atom: 2000, page size: 100000\n",
            "  master list distance cutoff = 14.5\n",
            "  ghost atom cutoff = 14.5\n",
            "  binsize = 7.25, bins = 9 9 11\n",
            "  3 neighbor lists, perpetual/occasional/extra = 3 0 0\n",
            "  (1) pair vexcluded, perpetual, half/full from (3)\n",
            "      attributes: half, newton on\n",
            "      pair build: halffull/newton\n",
            "      stencil: none\n",
            "      bin: none\n",
            "  (2) fix backbone, perpetual, copy from (1)\n",
            "      attributes: half, newton on\n",
            "      pair build: copy\n",
            "      stencil: none\n",
            "      bin: none\n",
            "  (3) fix backbone, perpetual\n",
            "      attributes: full, newton on\n",
            "      pair build: full/bin\n",
            "      stencil: full/bin/3d\n",
            "      bin: standard\n",
            "Setting up Verlet run ...\n",
            "  Unit style    : real\n",
            "  Current step  : 0\n",
            "  Time step     : 2\n",
            "WARNING: Communication cutoff 14.5 is shorter than a bond length based estimate of 15.655. This may lead to errors. (../comm.cpp:686)\n",
            "Per MPI rank memory allocation (min/avg/max) = 10.29 | 10.29 | 10.29 Mbytes\n",
            "Step Temp E_pair E_mol TotEng Press Volume \n",
            "       0          300    490.06176    59911.803    61069.889   -26349.405    295635.93 \n",
            "Loop time of 1.19209e-06 on 1 procs for 0 steps with 1722 atoms\n",
            "\n",
            "167.8% CPU use with 1 MPI tasks x no OpenMP threads\n",
            "\n",
            "MPI task timing breakdown:\n",
            "Section |  min time  |  avg time  |  max time  |%varavg| %total\n",
            "---------------------------------------------------------------\n",
            "Pair    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Bond    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Neigh   | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Comm    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Output  | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Modify  | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Other   |            | 1.192e-06  |            |       |100.00\n",
            "\n",
            "Nlocal:    1722 ave 1722 max 1722 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "Nghost:    0 ave 0 max 0 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "Neighs:    112733 ave 112733 max 112733 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "FullNghs:  225466 ave 225466 max 225466 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "\n",
            "Total # of neighbors = 225466\n",
            "Ave neighs/atom = 130.933\n",
            "Ave special neighs/atom = 2.65738\n",
            "Neighbor list builds = 0\n",
            "Dangerous builds = 0\n",
            "Total wall time: 0:00:02\n",
            "-----------------------------Reorganization-----------------------------\n",
            "\n",
            "\n",
            "****Storage information****\n",
            "The frustration data was stored in /home/ceramirez/github/frustrapy/Results_example/1fhj.done/FrustrationData\n",
            "-----------------------------Filtering-----------------------------\n",
            "/home/ceramirez/github/frustrapy/Results_example/2dn1.done/2dn1.pdb\n",
            "2dn1.pdb\n",
            "/home/ceramirez/github/frustrapy/Results_example/2dn1.done/2dn1.pdb_equivalences.txt\n",
            "HI\n",
            "/home/ceramirez/github/frustrapy/Results_example/2dn1.done/2dn1.pdb\n",
            "2dn1.pdb\n",
            "/home/ceramirez/github/frustrapy/Results_example/2dn1.done/2dn1.pdb_equivalences.txt\n",
            "-----------------------------Preparing files-----------------------------\n",
            "Pdb file: 2dn1\n",
            "Output file: 2dn1\n",
            "1 HIS missing CB atom!\n",
            "Abort!\n",
            "-----------------------------Setting options-----------------------------\n",
            "-----------------------------Calculating-----------------------------\n",
            "LAMMPS (3 Mar 2020)\n",
            "Reading data file ...\n",
            "  orthogonal box = (-20000 -20000 -20000) to (20000 20000 20000)\n",
            "  1 by 1 by 1 MPI processor grid\n",
            "  reading atoms ...\n",
            "  420 atoms\n",
            "  scanning bonds ...\n",
            "  3 = max bonds/atom\n",
            "  reading bonds ...\n",
            "  558 bonds\n",
            "Finding 1-2 1-3 1-4 neighbors ...\n",
            "  special bond factors lj:   0          1          1         \n",
            "  special bond factors coul: 0          1          1         \n",
            "  5 = max # of 1-2 neighbors\n",
            "  5 = max # of special neighbors\n",
            "  special bonds CPU = 0.000124931 secs\n",
            "  read_data CPU = 0.00362492 secs\n",
            "140 atoms in group alpha_carbons\n",
            "140 atoms in group beta_atoms\n",
            "140 atoms in group oxygens\n",
            "ABC flag on\n",
            "Water flag on\n",
            "Burial flag on\n",
            "Tertiary_Frustratometer flag on\n",
            "\n",
            "Neighbor list info ...\n",
            "  update every 1 steps, delay 5 steps, check yes\n",
            "  max neighbors/atom: 2000, page size: 100000\n",
            "  master list distance cutoff = 14.5\n",
            "  ghost atom cutoff = 14.5\n",
            "  binsize = 7.25, bins = 6 7 6\n",
            "  3 neighbor lists, perpetual/occasional/extra = 3 0 0\n",
            "  (1) pair vexcluded, perpetual, half/full from (3)\n",
            "      attributes: half, newton on\n",
            "      pair build: halffull/newton\n",
            "      stencil: none\n",
            "      bin: none\n",
            "  (2) fix backbone, perpetual, copy from (1)\n",
            "      attributes: half, newton on\n",
            "      pair build: copy\n",
            "      stencil: none\n",
            "      bin: none\n",
            "  (3) fix backbone, perpetual\n",
            "      attributes: full, newton on\n",
            "      pair build: full/bin\n",
            "      stencil: full/bin/3d\n",
            "      bin: standard\n",
            "Setting up Verlet run ...\n",
            "  Unit style    : real\n",
            "  Current step  : 0\n",
            "  Time step     : 2\n",
            "WARNING: Communication cutoff 14.5 is shorter than a bond length based estimate of 15.655. This may lead to errors. (../comm.cpp:686)\n",
            "Per MPI rank memory allocation (min/avg/max) = 8.605 | 8.605 | 8.605 Mbytes\n",
            "Step Temp E_pair E_mol TotEng Press Volume \n",
            "       0          300    160.04906    3815.2994    4145.9534    -13005.36    80213.491 \n",
            "Loop time of 9.53674e-07 on 1 procs for 0 steps with 420 atoms\n",
            "\n",
            "209.7% CPU use with 1 MPI tasks x no OpenMP threads\n",
            "\n",
            "MPI task timing breakdown:\n",
            "Section |  min time  |  avg time  |  max time  |%varavg| %total\n",
            "---------------------------------------------------------------\n",
            "Pair    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Bond    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Neigh   | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Comm    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Output  | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Modify  | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Other   |            | 9.537e-07  |            |       |100.00\n",
            "\n",
            "Nlocal:    420 ave 420 max 420 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "Nghost:    0 ave 0 max 0 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "Neighs:    23186 ave 23186 max 23186 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "FullNghs:  46372 ave 46372 max 46372 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "\n",
            "Total # of neighbors = 46372\n",
            "Ave neighs/atom = 110.41\n",
            "Ave special neighs/atom = 2.65714\n",
            "Neighbor list builds = 0\n",
            "Dangerous builds = 0\n",
            "Total wall time: 0:00:00\n",
            "-----------------------------Reorganization-----------------------------\n",
            "\n",
            "\n",
            "****Storage information****\n",
            "The frustration data was stored in /home/ceramirez/github/frustrapy/Results_example/2dn1.done/FrustrationData\n",
            "-----------------------------Filtering-----------------------------\n",
            "/home/ceramirez/github/frustrapy/Results_example/1m6k.done/1m6k.pdb\n",
            "1m6k.pdb\n",
            "/home/ceramirez/github/frustrapy/Results_example/1m6k.done/1m6k.pdb_equivalences.txt\n",
            "HI\n",
            "/home/ceramirez/github/frustrapy/Results_example/1m6k.done/1m6k.pdb\n",
            "1m6k.pdb\n",
            "/home/ceramirez/github/frustrapy/Results_example/1m6k.done/1m6k.pdb_equivalences.txt\n",
            "-----------------------------Preparing files-----------------------------\n",
            "Pdb file: 1m6k\n",
            "Output file: 1m6k\n",
            "1 SER missing CB atom!\n",
            "Abort!\n",
            "-----------------------------Setting options-----------------------------\n",
            "-----------------------------Calculating-----------------------------\n",
            "LAMMPS (3 Mar 2020)\n",
            "Reading data file ...\n",
            "  orthogonal box = (-20000 -20000 -20000) to (20000 20000 20000)\n",
            "  1 by 1 by 1 MPI processor grid\n",
            "  reading atoms ...\n",
            "  747 atoms\n",
            "  scanning bonds ...\n",
            "  3 = max bonds/atom\n",
            "  reading bonds ...\n",
            "  994 bonds\n",
            "Finding 1-2 1-3 1-4 neighbors ...\n",
            "  special bond factors lj:   0          1          1         \n",
            "  special bond factors coul: 0          1          1         \n",
            "  5 = max # of 1-2 neighbors\n",
            "  5 = max # of special neighbors\n",
            "  special bonds CPU = 0.000163078 secs\n",
            "  read_data CPU = 0.00339603 secs\n",
            "249 atoms in group alpha_carbons\n",
            "249 atoms in group beta_atoms\n",
            "249 atoms in group oxygens\n",
            "ABC flag on\n",
            "Water flag on\n",
            "Burial flag on\n",
            "Tertiary_Frustratometer flag on\n",
            "\n",
            "Neighbor list info ...\n",
            "  update every 1 steps, delay 5 steps, check yes\n",
            "  max neighbors/atom: 2000, page size: 100000\n",
            "  master list distance cutoff = 14.5\n",
            "  ghost atom cutoff = 14.5\n",
            "  binsize = 7.25, bins = 7 8 8\n",
            "  3 neighbor lists, perpetual/occasional/extra = 3 0 0\n",
            "  (1) pair vexcluded, perpetual, half/full from (3)\n",
            "      attributes: half, newton on\n",
            "      pair build: halffull/newton\n",
            "      stencil: none\n",
            "      bin: none\n",
            "  (2) fix backbone, perpetual, copy from (1)\n",
            "      attributes: half, newton on\n",
            "      pair build: copy\n",
            "      stencil: none\n",
            "      bin: none\n",
            "  (3) fix backbone, perpetual\n",
            "      attributes: full, newton on\n",
            "      pair build: full/bin\n",
            "      stencil: full/bin/3d\n",
            "      bin: standard\n",
            "Setting up Verlet run ...\n",
            "  Unit style    : real\n",
            "  Current step  : 0\n",
            "  Time step     : 2\n",
            "WARNING: Communication cutoff 14.5 is shorter than a bond length based estimate of 15.655. This may lead to errors. (../comm.cpp:686)\n",
            "Per MPI rank memory allocation (min/avg/max) = 8.646 | 8.646 | 8.646 Mbytes\n",
            "Step Temp E_pair E_mol TotEng Press Volume \n",
            "       0          300    405.61433    61919.693    62623.439   -41807.019     130227.4 \n",
            "Loop time of 1.19209e-06 on 1 procs for 0 steps with 747 atoms\n",
            "\n",
            "167.8% CPU use with 1 MPI tasks x no OpenMP threads\n",
            "\n",
            "MPI task timing breakdown:\n",
            "Section |  min time  |  avg time  |  max time  |%varavg| %total\n",
            "---------------------------------------------------------------\n",
            "Pair    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Bond    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Neigh   | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Comm    | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Output  | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Modify  | 0          | 0          | 0          |   0.0 |  0.00\n",
            "Other   |            | 1.192e-06  |            |       |100.00\n",
            "\n",
            "Nlocal:    747 ave 747 max 747 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "Nghost:    0 ave 0 max 0 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "Neighs:    48001 ave 48001 max 48001 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "FullNghs:  96002 ave 96002 max 96002 min\n",
            "Histogram: 1 0 0 0 0 0 0 0 0 0\n",
            "\n",
            "Total # of neighbors = 96002\n",
            "Ave neighs/atom = 128.517\n",
            "Ave special neighs/atom = 2.66131\n",
            "Neighbor list builds = 0\n",
            "Dangerous builds = 0\n",
            "Total wall time: 0:00:00\n",
            "-----------------------------Reorganization-----------------------------\n",
            "\n",
            "\n",
            "****Storage information****\n",
            "The frustration data was stored in /home/ceramirez/github/frustrapy/Results_example/1m6k.done/FrustrationData\n",
            "\n",
            "\n",
            "****Storage information****\n",
            "Frustration data for all Pdb's directory /home/ceramirez/github/frustrapy are stored in /home/ceramirez/github/frustrapy/Results_example/\n"
          ]
        }
      ],
      "source": [
        "# @title Frustratometer in Python\n",
        "mode = \"singleresidue\" # @param [\"configurational\", \"singleresidue\", \"mutational\"]\n",
        "pdbs_dir = \"/content\" # @param {type:\"string\"}\n",
        "results_dir = \"/home/ceramirez/github/frustrapy/Results_example\" # @param {type:\"string\"}\n",
        "example = True # @param {type:\"boolean\"}\n",
        "overwrite = True # @param {type:\"boolean\"}\n",
        "debug = False # @param {type:\"boolean\"}\n",
        "\n",
        "import sys\n",
        "# /home/ceramirez/github/frustrapy/frustrapy/functions.py\n",
        "# Import the functions from the frustrapy package\n",
        "import frustrapy.functions as frustrapy_functions\n",
        "\n",
        "# If the example is True, we will download the example files\n",
        "if example:\n",
        "    !wget -q http://www.rcsb.org/pdb/files/1fhj.pdb -O 1fhj.pdb\n",
        "    !wget -q http://www.rcsb.org/pdb/files/2dn1.pdb -O 2dn1.pdb\n",
        "    !wget -q http://www.rcsb.org/pdb/files/1m6k.pdb -O 1m6k.pdb\n",
        "\n",
        "    pdbs_dir = \"/home/ceramirez/github/frustrapy\"\n",
        "    results_dir = \"/home/ceramirez/github/frustrapy/Results_example\"\n",
        "    # Remove any previous results\n",
        "    !rm -rf /home/ceramirez/github/frustrapy/Results_example/*\n",
        "\n",
        "if overwrite:\n",
        "    if example:\n",
        "        !rm -rf /home/ceramirez/github/frustrapy/Results_example/*\n",
        "    else:\n",
        "        import os\n",
        "        # Convert the results_dir to an absolute path\n",
        "        results_dir = os.path.abspath(results_dir)\n",
        "        os.system(f\"rm -rf {results_dir}/*\")\n",
        "\n",
        "plots_dir_dict = frustrapy_functions.dir_frustration(\n",
        "    pdbs_dir=pdbs_dir,\n",
        "    mode=mode,\n",
        "    results_dir=results_dir,\n",
        "    debug=debug\n",
        ")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Plots"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "{'1fhj': {}, '2dn1': {}, '1m6k': {}}\n"
          ]
        }
      ],
      "source": [
        "print(plots_dir_dict)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "ename": "ValueError",
          "evalue": "Mime type rendering requires nbformat>=4.2.0 but it is not installed",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[18], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m plot \u001b[38;5;129;01min\u001b[39;00m plots_dir_dict[pdb]\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m      4\u001b[0m     fig \u001b[38;5;241m=\u001b[39m plots_dir_dict[pdb][plot]\n\u001b[0;32m----> 5\u001b[0m     \u001b[43mfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshow\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/Documents/Binaries/localcolabfold/conda/lib/python3.10/site-packages/plotly/basedatatypes.py:3410\u001b[0m, in \u001b[0;36mBaseFigure.show\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   3377\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   3378\u001b[0m \u001b[38;5;124;03mShow a figure using either the default renderer(s) or the renderer(s)\u001b[39;00m\n\u001b[1;32m   3379\u001b[0m \u001b[38;5;124;03mspecified by the renderer argument\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3406\u001b[0m \u001b[38;5;124;03mNone\u001b[39;00m\n\u001b[1;32m   3407\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   3408\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mplotly\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mio\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpio\u001b[39;00m\n\u001b[0;32m-> 3410\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mpio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshow\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
            "File \u001b[0;32m~/Documents/Binaries/localcolabfold/conda/lib/python3.10/site-packages/plotly/io/_renderers.py:394\u001b[0m, in \u001b[0;36mshow\u001b[0;34m(fig, renderer, validate, **kwargs)\u001b[0m\n\u001b[1;32m    389\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    390\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMime type rendering requires ipython but it is not installed\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    391\u001b[0m         )\n\u001b[1;32m    393\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m nbformat \u001b[38;5;129;01mor\u001b[39;00m Version(nbformat\u001b[38;5;241m.\u001b[39m__version__) \u001b[38;5;241m<\u001b[39m Version(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m4.2.0\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[0;32m--> 394\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    395\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMime type rendering requires nbformat>=4.2.0 but it is not installed\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    396\u001b[0m         )\n\u001b[1;32m    398\u001b[0m     ipython_display\u001b[38;5;241m.\u001b[39mdisplay(bundle, raw\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    400\u001b[0m \u001b[38;5;66;03m# external renderers\u001b[39;00m\n",
            "\u001b[0;31mValueError\u001b[0m: Mime type rendering requires nbformat>=4.2.0 but it is not installed"
          ]
        }
      ],
      "source": [
        "# Show the fg objects \n",
        "for pdb in plots_dir_dict.keys():\n",
        "    for plot in plots_dir_dict[pdb].keys():\n",
        "        fig = plots_dir_dict[pdb][plot]\n",
        "        fig.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Note: you may need to restart the kernel to use updated packages.\n"
          ]
        }
      ],
      "source": [
        "%pip install nbformat>=4.2.0"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Types of frustration modes the Frustratometer you can calculate in this notebook:\n",
        "\n",
        "| Frustration Mode | Description | Mathematical Formula | Example Calculation |\n",
        "|------------------|-------------|----------------------|---------------------|\n",
        "| Configurational  | Compares the native energy of each contact in the protein to a set of decoy energies from random variants of the same contact. A contact is considered frustrated if its native energy is higher than the average of the decoys. | $F_c = \\frac{E_n - \\langle E_d \\rangle}{\\sigma_d}$ <br><br> $E_n$ = native energy of contact <br> $\\langle E_d \\rangle$ = mean energy of decoys <br> $\\sigma_d$ = standard deviation of decoy energies | Native contact energy $E_n = -2.5$ <br> Mean decoy energy $\\langle E_d \\rangle = -5.2$ <br> Decoy std dev $\\sigma_d = 1.8$ <br><br> $F_c = \\frac{-2.5 - (-5.2)}{1.8} = 1.5$ <br><br> $F_c > 0$, so contact is frustrated |\n",
        "| Mutational       | Compares the native energy of each contact to the average energy of all possible mutations of the amino acids forming that contact. A contact is considered frustrated if mutating it makes the energy more favorable on average. | $F_m = \\frac{E_n - \\langle E_m \\rangle}{\\sigma_m}$ <br><br> $E_n$ = native energy of contact <br> $\\langle E_m \\rangle$ = mean energy of all mutations <br> $\\sigma_m$ = standard deviation of mutation energies | Native contact energy $E_n = -4.2$ <br> Mean mutation energy $\\langle E_m \\rangle = -6.8$ <br> Mutation std dev $\\sigma_m = 2.1$ <br><br> $F_m = \\frac{-4.2 - (-6.8)}{2.1} = 1.2$ <br><br> $F_m > 0$, so contact is frustrated |  \n",
        "| Single Residue   | Calculates the total frustration of all contacts a single residue is involved in. Residues with many frustrated contacts are considered highly frustrated. | $F_r = \\frac{1}{N} \\sum_{i=1}^N F_{c,i}$ <br><br> $F_{c,i}$ = configurational frustration of $i$th contact <br> $N$ = total number of contacts residue is involved in | Residue involved in 3 contacts: <br> $F_{c,1} = 0.8$ <br> $F_{c,2} = 1.2$ <br> $F_{c,3} = -0.5$ <br><br> $F_r = \\frac{1}{3}(0.8 + 1.2 + -0.5) = 0.5$ <br><br> $F_r > 0$, so residue is net frustrated |\n",
        "\n",
        "In a nuthshell:\n",
        "- Configurational frustration compares native contact energy to decoys\n",
        "- Mutational frustration compares native contact energy to average mutation energy  \n",
        "- Single residue frustration averages configurational frustration over all of a residue's contacts\n",
        "\n",
        "The key equations are:\n",
        "\n",
        "$F_c = \\frac{E_n - \\langle E_d \\rangle}{\\sigma_d}$ (configurational)\n",
        "\n",
        "$F_m = \\frac{E_n - \\langle E_m \\rangle}{\\sigma_m}$ (mutational)  \n",
        "\n",
        "$F_r = \\frac{1}{N} \\sum_{i=1}^N F_{c,i}$ (single residue)\n",
        "\n",
        "Where $E_n$ is the native energy, $\\langle E_d \\rangle$ and $\\langle E_m \\rangle$ are mean decoy and mutation energies, and $\\sigma_d$ and $\\sigma_m$ are the standard deviations of the decoy and mutation energy distributions.\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.12"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
